src/arch/arm/config.cmake

```diff
@@ -32,6 +32,9 @@ elseif(KernelArmCortexA72)
     # (https://developer.arm.com/documentation/100095/0001/memory-management-unit/about-the-mmu)
     set(KernelArmPASizeBits44 ON)
     math(EXPR KernelPaddrUserTop "(1 << 44)")
+elseif(KernelArmMorello)
+    set(KernelArmPASizeBits48 ON)
```

Comment:
Seems `KernelArmPASizeBits48` is a new CMake variable, but it's not used anywhere at the moment. Can't we leave this at 44 bit for now to keep things simple and add the rest later if this is needed?

Reply:
If we punt on the address space size, we can as well pretend it's a KernelArmCortexA72 and make it even simpler.

Reply:
I'd rather make naming/numbers correct from the start if they're not causing issues with verification, functional correctness, or affecting other platforms. Simplicity is good, but in this case it would be misleading. Further, it's just a few lines. This will also save us renaming later when/if we add more Morello features (e.g., CHERI, hypervisor support, etc). 

Regarding the size, it's used to calculate the `USER_TOP`.  Even though `KernelArmPASizeBits48` is currently unused, the fact is Morello has 48 PA and it's not going to be misleading for others who are reading this to think that Morello only supports 44 PA. So, if we don't want to define `KernelArmPASizeBits48` now, I'd just delete it but not make it 44.
Related to that, I don't remember why the full 48 PA size isn't allowed in non hypervisor config? I could only have 47 due to [this assert](https://github.com/seL4/seL4/blob/ed613d7d490be91f7f164ff98e09d9098de64d6e/include/arch/arm/arch/64/mode/hardware.h#L222)? AFAIK, canonical PPTRs are virtual, but this is a 48-bit physical address

Reply:
> Related to that, I don't remember why the full 48 PA size isn't allowed in non hypervisor config? I could only have 47 due to [this assert](https://github.com/seL4/seL4/blob/ed613d7d490be91f7f164ff98e09d9098de64d6e/include/arch/arm/arch/64/mode/hardware.h#L222)? AFAIK, canonical PPTRs are virtual, but this is a 48-bit physical address

Device untyped caps are created for the full physical address space. Device untyped caps are created using generated bitfield accessor functions. These asserts were needed to avoid asserts in the bitfield generator accessor functions during boot.  You could try removing that assert and see if there's bitfield asserts during boot (in debug build).

Reply:
Yeah, when I delete this assert, there's another one from the bitfield generator that fails after the root task starts. Is that something expected or is it an issue to do with the bitfield generator? This cuts the device untype range to half to the potential physical address space on Morello

Reply:
It's an issue with the size of the field it's writing to. The bitfield generator just checks that you're not trying to store a value large than what fits into the field. I'm not sure which field exactly it's failing for, but it seems that it would need more bit width to store the value that is provided.

Reply:
@lsf37 Submitted https://github.com/seL4/seL4/pull/1163 for that, but it's not a blocker for this PR as it can just proceed with 47-bit PA for now

Reply:
`capPtr` on device untypeds is absolutely used, so #1163 definitely is not possible. They are not dereferenced for device memory, but they are used esp for comparison with other kernel virtual pointers and need to satisfy the same basic conditions as other kernel virtual pointers. You will actually get wrong results if they don't.

Reply:
The only way I can see to relax that would be to find all places where that happens and to make special cases, comparing physical instead of virtual addresses in those case. I *think* that might work, but there are a lot of assumptions in how the CDT is laid out and how pointers relate to each other, so I might be wrong. In any case, that would need verification funding to do.

Reply:
One of the assumptions that the 64-bit kernel design currently makes is that the kernel window is large enough to hold all of addressable physical memory. With 48 bits of addressable virtual memory, 47-bit PA limit is probably the larges as there needs to be parts of the kernel virtual address space that isn't used for the direct mapping window. Getting larger than 47-bit would probably require a different approach for the kernel window mapping.

Reply:
In hyp mode, the kernel uses a separate page table for the kernel mappings. @kent-mcleod I think you mentioned a while ago that armv8 would support that in non-hyp mode as well. That might solve at least some of the problem, and since we're only verifying hyp mode at the moment, it would not have any verification impact.

There is still the problem that addresses might overflow 48 bits in the translation between kernel virtual and physical, but that could probably at least affect a smaller region of memory, not the full top half of the physical address space.

Reply:
@lsf37 Thanks for the clarification. The Morello port could go ahead with just 47-bit PA as per this PR then, and we could create an issue for future 48-bit PA support in seL4 in general